session.go

```diff
@@ -4956,6 +4988,140 @@ func (r *writeCmdResult) BulkErrorCases() []BulkErrorCase {
 	return ecases
 }
 
+func (c *Collection) writeOpWithOpMsg(socket *mongoSocket, serverInfo *mongoServerInfo, op interface{}, ordered, bypassValidation bool, safeOp *queryOp) (*LastError, error) {
+	var cmd bson.D
+	var documents []interface{}
+	var docSeqID string
+	canUseOpMsg := true
+	switch msgOp := op.(type) {
+	case *insertOp:
+		cmd = bson.D{
+			{Name: "insert", Value: c.Name},
+		}
+		docSeqID = "documents"
+		documents = msgOp.documents
+	case bulkUpdateOp:
+		cmd = bson.D{
+			{Name: "update", Value: c.Name},
+		}
+		docSeqID = "updates"
+		documents = msgOp
+	case bulkDeleteOp:
+		cmd = bson.D{
+			{Name: "delete", Value: c.Name},
+		}
+		docSeqID = "deletes"
+		documents = msgOp
+	default:
+		canUseOpMsg = false
+	}
+
+	if canUseOpMsg {
+		//msg flags, see https://docs.mongodb.com/master/reference/mongodb-wire-protocol/#flag-bits
+		flags := uint32(0)
+
+		var writeConcern interface{}
+		if safeOp == nil {
+			// unacknowledged writes
+			flags |= opMsgFlagMoreToCome
+			writeConcern = bson.D{{Name: "w", Value: 0}}
+		} else {
+			writeConcern = safeOp.query.(*getLastError)
+		}
+
+		cmd = append(cmd, bson.DocElem{
+			Name: "$db", Value: c.Database.Name,
+		}, bson.DocElem{
+			Name: "ordered", Value: ordered,
+		}, bson.DocElem{
+			Name: "writeConcern", Value: writeConcern,
+		}, bson.DocElem{
+			Name: "bypassDocumentValidation", Value: bypassValidation,
+		})
+
+		body := msgSection{
+			payloadType: msgPayload0,
+			data:        cmd,
+		}
+
+		n := 0
+		modified := 0
+		var errs []BulkErrorCase
+		var lerr LastError
+
+		l := len(documents)
+		batchNb := (l / serverInfo.MaxWriteBatchSize) + 1
+		if l != 0 && (l%serverInfo.MaxWriteBatchSize) == 0 {
+			batchNb--
+		}
+		count := 0
+
+		for count < batchNb {
+			start := count * serverInfo.MaxWriteBatchSize
+			length := l - start
+			if length > serverInfo.MaxWriteBatchSize {
+				length = serverInfo.MaxWriteBatchSize
```

Comment:
I know this _should_ never happen, but I think we should stop and return an error if it does - silently dropping data would be surprising to the caller.

Reply:
Actually, data is not silently dropped here: if we have to insert/update/delete more than `MaxWriteBatchSize` documents, the driver split the list of documents in batch of `MaxWriteBatchSize` and send them to the database one by one (same behavior in the old wire protocol)